src/scheduler/replay.rs

```diff
@@ -64,26 +73,46 @@ impl Scheduler for ReplayScheduler {
         }
     }
 
-    fn next_task(&mut self, runnable: &[TaskId], _current: Option<TaskId>, _is_yielding: bool) -> Option<TaskId> {
-        if self.steps >= self.schedule.steps.len() {
-            assert!(self.allow_incomplete, "schedule ended early");
-            return None;
-        }
-        match self.schedule.steps[self.steps] {
-            ScheduleStep::Random => {
-                panic!("expected context switch but next schedule step is random choice");
+    fn next_task(&mut self, runnable: &[&Task], _current: Option<TaskId>, _is_yielding: bool) -> Option<TaskId> {
+        loop {
+            if self.steps >= self.schedule.steps.len() {
+                assert!(self.allow_incomplete, "schedule ended early");
+                return None;
             }
-            ScheduleStep::Task(next) => {
-                if !runnable.contains(&next) {
-                    assert!(
-                        self.allow_incomplete,
-                        "scheduled task is not runnable, expected to run {:?}, but choices were {:?}",
-                        next, runnable
-                    );
-                    None
-                } else {
-                    self.steps += 1;
-                    Some(next)
+            match self.schedule.steps[self.steps] {
+                ScheduleStep::Random => {
+                    panic!("expected context switch but next schedule step is random choice");
+                }
+                ScheduleStep::Task(next) => {
+                    if let Some(task) = runnable.iter().find(|t| t.id() == next) {
+                        self.steps += 1;
+                        if let Some(target_clock) = &self.target_clock {
+                            if task.clock <= *target_clock {
+                                // The target event causally depends on this
+                                // event, so we schedule it.
+                                return Some(next);
+                            } else {
+                                // The target even is concurrent with this
+                                // event, so it is irrelevant to the replay.
+                                // At this point, we also need to skip over
+                                // any random steps made by the thread that
+                                // would have been scheduled.
+                                while let Some(ScheduleStep::Random) = self.schedule.steps.get(self.steps) {
```

Comment:
Would be good to generate a `trace!` here recording that a step was skipped (and also how many random steps).  Also add a counter to the scheduler so we can also report how many steps were skipped during a replay run.